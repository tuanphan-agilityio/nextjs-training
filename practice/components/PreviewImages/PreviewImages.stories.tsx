import { Story, Meta } from '@storybook/react';

import PreviewImages, { PreviewImagesProps } from '.';

export default {
  title: 'PreviewImages',
  component: PreviewImages,
  decorators: [
    (Story) => (
      <div className='max-w-[588px]'>
        <Story />
      </div>
    ),
  ],
} as Meta;

const Template: Story<PreviewImagesProps> = (args) => (
  <PreviewImages {...args} />
);

export const Default = Template.bind({});
Default.args = {
  imageHrefs: [
    '/images/product-1.jpg',
    '/images/product-2.jpg',
    '/images/product-3.jpg',
  ],
};
