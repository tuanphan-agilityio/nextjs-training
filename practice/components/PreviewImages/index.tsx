import Image from 'next/image';
import { FC, memo, useState, useCallback } from 'react';

interface PreviewImagesProps {
  imageHrefs: string[];
}

const PreviewImages: FC<PreviewImagesProps> = ({ imageHrefs }) => {
  const [currentSlideIdx, setCurrentSlideIdx] = useState(0);

  const handleImageClick = useCallback((index: number) => {
    setCurrentSlideIdx(index);
  }, []);

  const handleImageClickWithIndex = useCallback(
    (index: number) => () => {
      handleImageClick(index);
    },
    [handleImageClick],
  );

  const renderImage = (imageHref: string, index: number) => (
    <Image
      key={index}
      src={imageHref}
      width={180}
      height={180}
      alt='Product'
      onClick={handleImageClickWithIndex(index)}
      objectFit='cover'
      className='hover:cursor-pointer'
      data-testid={`image-${index}`}
    />
  );

  return (
    <div className='flex flex-col gap-6'>
      <Image
        src={imageHrefs[currentSlideIdx]}
        layout='responsive'
        width={588}
        height={400}
        alt='Product'
        objectFit='cover'
        data-testid={`main-image-${currentSlideIdx}`}
        priority
      />

      <div className='flex gap-6 w-full overflow-auto'>
        {imageHrefs.map(renderImage)}
      </div>
    </div>
  );
};

export type { PreviewImagesProps };

export default memo(PreviewImages);
