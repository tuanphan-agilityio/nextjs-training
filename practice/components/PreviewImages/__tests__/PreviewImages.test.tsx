import { render, screen } from '@testing-library/react';

import PreviewImages, { PreviewImagesProps } from '..';

describe('PreviewImages Component', () => {
  const mockProps: PreviewImagesProps = {
    imageHrefs: [
      '/images/product-1.webp',
      '/images/product-2.webp',
      '/images/product-3.webp',
    ],
  };

  it('renders the first image as the initial main image', () => {
    render(<PreviewImages {...mockProps} />);
    const mainImage = screen.getByTestId(/main-image-\d/);
    expect(mainImage).toHaveAttribute('data-testid', 'main-image-0');
  });

  it('renders all thumbnails', () => {
    render(<PreviewImages {...mockProps} />);
    const thumbnails = screen.getAllByTestId(/image-\d/);
    expect(thumbnails).toHaveLength(mockProps.imageHrefs.length + 1);
  });

  it('matches the snapshot', () => {
    const { container } = render(<PreviewImages {...mockProps} />);
    expect(container).toMatchSnapshot();
  });
});
