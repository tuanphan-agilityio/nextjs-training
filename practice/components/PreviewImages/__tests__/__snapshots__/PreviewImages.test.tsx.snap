// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PreviewImages Component matches the snapshot 1`] = `
<div>
  <div
    class="flex flex-col gap-6"
  >
    <div
      style="overflow: hidden; box-sizing: border-box; display: inline-block; position: relative; width: 588px; height: 400px;"
    >
      <noscript />
      <img
        alt="Product"
        data-testid="main-image-0"
        decoding="async"
        src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
        style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: cover;"
      />
    </div>
    <div
      class="flex gap-6 w-full overflow-auto"
    >
      <div
        style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
      >
        <div
          style="box-sizing: border-box; display: block; max-width: 100%;"
        >
          <img
            alt=""
            aria-hidden="true"
            role="presentation"
            src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTgwIiBoZWlnaHQ9IjE4MCIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB2ZXJzaW9uPSIxLjEiLz4="
            style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
          />
        </div>
        <noscript />
        <img
          alt="Product"
          class="hover:cursor-pointer"
          data-testid="image-0"
          decoding="async"
          src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
          style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: cover;"
        />
      </div>
      <div
        style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
      >
        <div
          style="box-sizing: border-box; display: block; max-width: 100%;"
        >
          <img
            alt=""
            aria-hidden="true"
            role="presentation"
            src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTgwIiBoZWlnaHQ9IjE4MCIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB2ZXJzaW9uPSIxLjEiLz4="
            style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
          />
        </div>
        <noscript />
        <img
          alt="Product"
          class="hover:cursor-pointer"
          data-testid="image-1"
          decoding="async"
          src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
          style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: cover;"
        />
      </div>
      <div
        style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
      >
        <div
          style="box-sizing: border-box; display: block; max-width: 100%;"
        >
          <img
            alt=""
            aria-hidden="true"
            role="presentation"
            src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTgwIiBoZWlnaHQ9IjE4MCIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB2ZXJzaW9uPSIxLjEiLz4="
            style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
          />
        </div>
        <noscript />
        <img
          alt="Product"
          class="hover:cursor-pointer"
          data-testid="image-2"
          decoding="async"
          src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
          style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: cover;"
        />
      </div>
    </div>
  </div>
</div>
`;
